client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java

```diff
@@ -155,6 +156,15 @@
     */
    NotifyingFuture<Boolean> replaceWithVersionAsync(K key, V newValue, long version, int lifespanSeconds, int maxIdleSeconds);
 
+   /**
+    * Retrieve entries from the server
+    *
+    * @param filterConverterFactory Factory name for the KeyValueFilterConverter or null
+    * @param segments               The segments to iterate on or null if all segments should be iterated
+    * @param batchSize              The number of entries transferred from the server at a time
+    * @return Iterator for the entries
+    */
+   CloseableIterator<Entry<Object, Object>> retrieveEntries(String filterConverterFactory, Set<Integer> segments, int batchSize);
```

Comment:
> AdvancedRemoteCache interface?

^ Not sure that's needed really. What you need is an extra operation client side, which given N keys, it can return you the set of segments for those keys, and then you can call this method that you've designed. If the client is running within Infinispan itself, it does not need to call that extra client side operation. 

The extra operation can probably be computed directly in the remote cache client since the client knows about all the segments, so given some keys, it should be able to give you the segments.